network/tcp.go

```diff
@@ -0,0 +1,422 @@
+		// read the size of the next packet
+		n, err := c.conn.Read(b)
+		// if error then quit
+		if err != nil {
+
+			return nil, handleError(err)
+		}
+		// put it in the longterm buffer
+		if _, err := buffer.Write(b[:n]); err != nil {
+			log.Error("Couldn't write to buffer:", err)
+		}
+		read += Size(n)
+		b = b[n:]
+	}
+
+	// set the size read
+	c.updateRx(uint64(read))
+	return buffer.Bytes(), nil
+}
+
+// how many bytes do we write at once on the socket
+// 1400 seems a safe choice regarding the size of a ethernet packet.
+// https://stackoverflow.com/questions/2613734/maximum-packet-size-for-a-tcp-connection
+const maxChunkSize Size = 1400
+
+// Send will convert the NetworkMessage into an ApplicationMessage
+// and send it with send()
+// Returns an error if anything was wrong
+func (c *TCPConn) Send(ctx context.Context, obj Body) error {
+	c.sendMutex.Lock()
+	defer c.sendMutex.Unlock()
+	am, err := NewNetworkPacket(obj)
+	if err != nil {
+		return fmt.Errorf("Error converting packet: %v\n", err)
+	}
+	log.Lvlf5("Message SEND => %+v", am)
+	var b []byte
+	b, err = am.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("Error marshaling  message: %s", err.Error())
+	}
+	return c.send(b)
+}
+
+// send takes care of sending this slice of bytes FULLY to the connection
+func (c *TCPConn) send(b []byte) error {
+	// First write the size
+	packetSize := Size(len(b))
+	if err := binary.Write(c.conn, globalOrder, packetSize); err != nil {
+		return err
+	}
+	// Then send everything through the connection
+	// Send chunk by chunk
+	var sent Size
+	for sent < packetSize {
+		length := packetSize - sent
+		if length > maxChunkSize {
+			length = maxChunkSize
+		}
+
+		// Sending 'length' bytes
+		log.Lvl4("Sending from", c.conn.LocalAddr(), "to", c.conn.RemoteAddr())
+		n, err := c.conn.Write(b[:length])
+		if err != nil {
+			log.Error("Couldn't write chunk starting at", sent, "size", length, err)
+			return handleError(err)
+		}
+		sent += Size(n)
+		log.Lvl5("Sent", sent, "out of", packetSize)
+
+		// bytes left to send
+		b = b[n:]
+	}
+	// update stats on the connection
+	c.updateTx(uint64(packetSize))
+	return nil
+}
+
+// Remote returns the name of the peer at the end point of
+// the connection
+func (c *TCPConn) Remote() Address {
+	return c.endpoint
+}
+
+// Local returns the local address and port
+func (c *TCPConn) Local() Address {
+	return NewTCPAddress(c.conn.LocalAddr().String())
+}
+
+// Type implements the Conn interface
+func (c *TCPConn) Type() ConnType {
+	return PlainTCP
+}
+
+// Close ... closes the connection
+func (c *TCPConn) Close() error {
+	c.closedMut.Lock()
+	defer c.closedMut.Unlock()
+	if c.closed == true {
+		return nil
+	}
+	err := c.conn.Close()
+	c.closed = true
+	if err != nil {
+		return handleError(err)
+	}
+	return nil
+}
+
+// handleError produces the higher layer error depending on the type
+// so user of the package can know what is the cause of the problem
+func handleError(err error) error {
```

Comment:
Not really because in golang-network errors are defined in here https://golang.org/pkg/net/#OpError . And it can contains any kind of errors so I tried to make the most common ones in variables so we can do equality like ` if err == ErrClosed` or `if err == ErrEOF` and such, instead of looking at the string the error outputs...